README.md
# ChooseALicense-cli

[![Latest Version][1]][2]
[![Build Status][3]][4]

Bring http://choosealicense.com to your terminal. Choose a license in your terminal.

![demo][]

## Install

    $ pip install choosealicense-cli

## Usage

List all the available licenses:

    $ license show

Show the info of the specified license:

    $ license info LICENSE(e.g. mit)

Show the default context for the license template:

    $ license context LICENSE(e.g. mit)

Generate a license:

    $ license generate LICENSE(e.g. mit)

Get detailed help message via `license -h` and `license show/info/context/generate -h`

## License

MIT.

[1]: http://img.shields.io/pypi/v/choosealicense-cli.svg
[2]: https://pypi.python.org/pypi/choosealicense-cli
[3]: https://travis-ci.org/lord63/choosealicense-cli.svg
[4]: https://travis-ci.org/lord63/choosealicense-cli
[demo]: https://cloud.githubusercontent.com/assets/5268051/7150903/b7f16168-e354-11e4-91b5-0965a86c8158.jpeg

ChangeLog
## 2020.04.10 v1.1.0
- fix error when git config has unicode, #6
- add support github token via env var, #9

## 2019.06.06 v1.0.0 [BREAK CHANGE]
- update to latest github api
- drop python 3.3 support and add 3.6 support

## 2016.03.19 v0.3.0

- show version info: -v -> -V
- python 3.5 support
- wheel support
- refactor some code
- add code coverage, style check when running the tests

## 2015.04.15 v0.2.0

- better exception handle
- add support for python 3
- add test

## 2015.04.09 v0.1.0

- feature supported:
    - can list all the license
    - can show the info of the license
    - show the context for the license template
    - generate a license


choosealicense/test/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from os import path
import os

import pytest
import responses
from click.testing import CliRunner

ROOT = path.join(path.abspath(path.dirname(__file__)), 'responses')


@pytest.yield_fixture
def mock_api():
    # Mock get the license list.
    with open(path.join(ROOT, 'licenses.json')) as f:
        mock_response_body = f.read()
    responses.add(responses.GET, 'https://api.github.com/licenses',
                  body=mock_response_body)

    # Mock get each license template.
    for license in os.listdir(path.join(ROOT, 'licenses')):
        with open(path.join(ROOT, 'licenses/{0}'.format(license))) as f:
            mock_response_body = f.read()
        responses.add(responses.GET,
                      'https://api.github.com/licenses/{0}'.format(
                          path.splitext(license)[0]),
                      body=mock_response_body, content_type='application/json')

    # Mock get the invalid license.
    with open(path.join(ROOT, 'not_found.json')) as f:
        mock_not_found_body = f.read()
    responses.add(responses.GET, 'https://api.github.com/licenses/invalid',
                  body=mock_not_found_body, content_type='application/json')

    responses.start()
    yield responses
    responses.stop()


@pytest.fixture(scope='function')
def runner(mock_api):
    return CliRunner()


choosealicense/test/test_info.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
    Tests for the `license info` function
"""

import pytest

from choosealicense.cli import info


@pytest.mark.usefixtures('mock_api')
class TestInfo():
    def test_show_license_info(self, runner):
        all_the_licenses = {
            'agpl-3.0': ('GNU Affero General Public License v3.0', 12),
            'apache-2.0': ('Apache License 2.0', 10),
            'bsd-2-clause': ('BSD 2-Clause "Simplified" License', 7),
            'bsd-3-clause': ('BSD 3-Clause "New" or "Revised" License', 7),
            'epl-2.0': ('Eclipse Public License 2.0', 10),
            'gpl-2.0': ('GNU General Public License v2.0', 10),
            'gpl-3.0': ('GNU General Public License v3.0', 11),
            'lgpl-2.1': ('GNU Lesser General Public License v2.1', 10),
            'lgpl-3.0': ('GNU Lesser General Public License v3.0', 11),
            'mit': ('MIT License', 7),
            'mpl-2.0': ('Mozilla Public License 2.0', 11),
            'unlicense': ('The Unlicense', 6)
        }

        for short_name, fullname_and_rules_number in all_the_licenses.items():
            result = runner.invoke(info, [short_name])
            output, exit_code = result.output, result.exit_code
            rules = output.split('Limitations\n')[1].split('\n')
            flat_rules = sum([item.split() for item in rules], [])
            fullname, rules_number = fullname_and_rules_number

            assert exit_code == 0
            assert '</a>' not in output
            assert fullname in output
            assert '{0:<25}{1:<25}{2}'.format(
                'Permissions', 'Conditions', 'Limitations') in output
            assert rules_number == len(flat_rules)

    def test_show_invalid_license_info(self, runner):
        result = runner.invoke(info, ['invalid'])
        output, exit_code = result.output, result.exit_code
        assert exit_code != 0
        assert output == ("Error: Invalid license name, use `license show` "
                          "to get the all available licenses.\n")


choosealicense/test/test_show.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
    Tests for the `license show` function
"""

import pytest

from choosealicense.cli import show


@pytest.mark.usefixtures('mock_api')
class TestShow():
    def test_show_all_the_licenses(self, runner):
        result = runner.invoke(show)
        it_should_be = ('agpl-3.0, apache-2.0, bsd-2-clause, '
                        'bsd-3-clause, epl-2.0, gpl-2.0, gpl-3.0, '
                        'lgpl-2.1, lgpl-3.0, mit, mpl-2.0, unlicense')
        assert result.exit_code == 0
        assert result.output.strip() == it_should_be


choosealicense/ui.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import

import re
import textwrap

from click import echo, secho
from click.termui import get_terminal_size


class InfoUI(object):
    def __init__(self, info):
        self.info = info

    def build_ui(self):
        """Build the CLI UI for `license info`."""
        self._print_name(self.info['name'])
        self._print_description(self.info['description'])
        self._print_rule_list(self.info['permissions'],
                              self.info['conditions'],
                              self.info['limitations'])

    def _print_name(self, name):
        """Print the license name in a nice way."""
        secho(name, bold=True)
        echo()

    def _print_description(self, text):
        """Print the short description for the license."""
        # Strip the link tag, but this solution is not elegant enough.
        tag_with_text_re = r'<a href=[^<]*</a>'
        text_in_tag_re = r'(?<=">).*(?=</a>)'
        for link in re.findall(tag_with_text_re, text):
            text = text.replace(link, re.search(text_in_tag_re, link).group())
        width, _ = get_terminal_size()
        width = width if width < 78 else 78
        for line in textwrap.wrap(text, width):
            echo(line)
        echo()

    def _print_rule_list(self, permissions, conditions, limitations):
        """Print the rule list like the way on the web."""
        max_rule_num = max(list(
            map(len, [permissions, conditions, limitations])
        ))
        for item in [permissions, conditions, limitations]:
            if len(item) < max_rule_num:
                item.extend(' '*(max_rule_num-len(item)))
        secho('{:<25}'.format('Permissions'), nl=False, bold=True, fg='green')
        secho('{:<25}'.format('Conditions'), nl=False, bold=True, fg='blue')
        secho('Limitations', bold=True, fg='red')
        for permision, condition, limitation in zip(
                permissions, conditions, limitations):
            secho('{:<25}'.format(permision), nl=False)
            secho('{:<25}'.format(condition), nl=False)
            secho(limitation)


choosealicense/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import

import re

import click
from click import echo

from choosealicense import __version__
from choosealicense.utils import get_default_context, get_api_response
from choosealicense.ui import InfoUI


LICENSE_WITH_CONTEXT = ['mit', 'artistic-2.0', 'bsd-2-clause',
                        'bsd-3-clause', 'isc']


@click.group(context_settings={'help_option_names': ('-h', '--help')})
@click.version_option(__version__, '-V', '--version', message='%(version)s')
def cli():
    """ChooseALicense in your terminal."""
    pass  # pragma: no cover


@cli.command()
def show():
    """List all the license."""
    response = get_api_response('https://api.github.com/licenses')
    keys = [item['key'] for item in response.json()]
    echo(', '.join(keys))


@cli.command()
@click.argument('license')
def info(license):
    """Show the info of the specified license."""
    response = get_api_response(
        'https://api.github.com/licenses/{0}'.format(license))
    try:
        InfoUI(response.json()).build_ui()
    except KeyError:
        raise click.ClickException("Invalid license name, use `license show` "
                                   "to get the all available licenses.")


@cli.command()
@click.option('--fullname', '-f', help="Copyright holder's name")
@click.option('--year', '-y', help="Copyright year")
@click.option('--email', '-e', help="Copyright holder's email")
@click.option('--project', '-p', help="The project organization")
@click.argument('license')
def generate(license, fullname, year, email, project):
    """Generate the specified license."""
    response = get_api_response(
        'https://api.github.com/licenses/{0}'.format(license))

    try:
        license_template = response.json()['body']
    except KeyError:
        raise click.ClickException("Invalid license name, use `license show` "
                                   "to get the all available licenses.")

    if license not in LICENSE_WITH_CONTEXT:
        echo(license_template, nl=False)
    else:
        context_variable = re.findall(r'\[(\w+)\]', response.json()['body'])
        raw_context = {'fullname': fullname, 'year': year, 'email': email,
                       'project': project}
        user_context = {key: value for (key, value) in raw_context.items()
                        if value is not None}
        context = get_default_context()
        context.update(user_context)
        for item in context_variable:
            license_template = license_template.replace('[{0}]'.format(item),
                                                        context[item])
        echo(license_template, nl=False)


@cli.command()
@click.argument('license')
def context(license):
    """Show the default context for the license."""
    if license not in LICENSE_WITH_CONTEXT:
        echo("Just use it, there's no context for the license.")
    else:
        response = get_api_response(
            'https://api.github.com/licenses/{0}'.format(license))
        context = re.findall(r'\[(\w+)\]', response.json()['body'])
        default_context = get_default_context()
        echo('The template has following defaults:')
        for item in context:
            echo(u'\t{0}: {1}'.format(item, default_context[item]))
        echo('You can overwrite them at your ease.')


choosealicense/utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import absolute_import

from datetime import date
import os
import subprocess
import time

import click
import requests


doc = 'https://github.com/lord63/choosealicense-cli/wiki/Use-a-github-token-to-get-a-higher-api-rate-limit'


def get_api_response(url):
    headers = {'accept': 'application/vnd.github.drax-preview+json'}
    github_token = os.environ.get('CHOOSEALICENSE_GITHUB_TOKEN', '')
    if github_token != '':
        headers.update({'Authorization': 'token {0}'.format(github_token)})
    try:
        r = requests.get(url, headers=headers)
    except requests.exceptions.RequestException as err:
        raise click.ClickException('send api request failed, error: {0}'.format(err))
    if r.status_code == 403 and r.headers.get('X-RateLimit-Remaining', '') == '0':
        wait_seconds = int(r.headers['X-RateLimit-Reset']) - int(time.time())
        if (wait_seconds) < 60:
            reset_at = '{0} seconds later'.format(wait_seconds)
        else:
            reset_at = '{0} minutes later'.format(wait_seconds / 60)
        raise click.ClickException(
                'Github api rate limit exceeded, limit will be reset {0}. \n'
                'Or you can use a github token to get a higher limit, see doc:\n'
                '{1}'.format(reset_at, doc)
        )
    return r


def _get_git_info(name):
    """Get git configuration info, e.g. user.email."""
    try:
        command_list = ('git config --get user.{0}'.format(name)).split()
        result = subprocess.check_output(command_list).strip().decode('utf-8')
    except subprocess.CalledProcessError:
        raise click.ClickException(
            "Please configure your git(user.{0}).".format(name))
    else:
        return result


def get_default_context():
    """Get the default context for the license template."""
    year = str(date.today().year)
    fullname = _get_git_info('name')
    email = _get_git_info('email')
    return {'year': year, 'fullname': fullname, 'email': email,
            'project': 'the copyright holder'}


choosealicense/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import textwrap
import re
from datetime import date
import subprocess

import requests
import click
from click import echo, secho
from click.termui import get_terminal_size

from choosealicense import __version__


LICENSE_WITH_CONTEXT = ['mit', 'artistic-2.0', 'bsd-2-clause',
                        'bsd-3-clause', 'isc']


def print_name(name):
    """Print the license name in a nice way."""
    secho(name, bold=True)
    secho('='*len(name), bold=True)
    echo()


def print_description(text):
    """Print the short description for the license."""
    width, _ = get_terminal_size()
    width = width if width < 78 else 78
    for line in textwrap.wrap(text, width):
        echo(line)
    echo()


def print_rule_list(required, permitted, forbidden):
    """Print the rule list like the way on the web."""
    max_rule_num = max(list(map(len, [required, permitted, forbidden])))
    for item in [required, permitted, forbidden]:
        if len(item) < max_rule_num:
            item.extend(' '*(max_rule_num-len(item)))
    secho('{:<25}'.format('Required'), nl=False, bold=True, fg='blue')
    secho('{:<25}'.format('Permitted'), nl=False, bold=True, fg='green')
    secho('Forbidden', bold=True, fg='red')
    for require, permit, forbid in zip(required, permitted, forbidden):
        secho('{:<25}'.format(require), nl=False)
        secho('{:<25}'.format(permit), nl=False)
        secho(forbid)


def get_default_context():
    """Get the default context for the license template."""
    year = str(date.today().year)
    try:
        fullname = subprocess.check_output(
            'git config --get user.name'.split()
        ).strip().decode('utf-8')
    except subprocess.CalledProcessError:
        raise click.ClickException("Please configure your git(user.name).")
    try:
        email = subprocess.check_output(
            'git config --get user.email'.split()
        ).strip().decode('utf-8')
    except subprocess.CalledProcessError:
        raise click.ClickException("Please configure your git(user.email).")
    return {'year': year, 'fullname': fullname, 'email': email,
            'project': 'the copyright holder'}


@click.group(context_settings={'help_option_names': ('-h', '--help')})
@click.version_option(__version__, '-v', '--version', message='%(version)s')
def cli():
    """ChooseALicense in your terminal."""
    pass


@cli.command()
def show():
    """List all the license."""
    response = requests.get(
        'https://api.github.com/licenses',
        headers={'accept': 'application/vnd.github.drax-preview+json'})
    keys = [item['key'] for item in response.json()]
    echo(', '.join(keys))


@cli.command()
@click.argument('license')
def info(license):
    """Show the info of the specified license."""
    response = requests.get(
        'https://api.github.com/licenses/{0}'.format(license),
        headers={'accept': 'application/vnd.github.drax-preview+json'})
    try:
        print_name(response.json()['name'])
        print_description(response.json()['description'])
        print_rule_list(response.json()['required'],
                        response.json()['permitted'],
                        response.json()['forbidden'])
    except KeyError:
        raise click.ClickException("Invalid license name, use `license show` "
                                   "to get the all available licenses.")


@cli.command()
@click.option('--fullname', '-f', help="Copyright holder's name")
@click.option('--year', '-y', help="Copyright year")
@click.option('--email', '-e', help="Copyright holder's email")
@click.option('--project', '-p', help="The project organization")
@click.argument('license')
def generate(license, fullname, year, email, project):
    """Generate the specified license."""
    response = requests.get(
        'https://api.github.com/licenses/{0}'.format(license),
        headers={'accept': 'application/vnd.github.drax-preview+json'})

    try:
        license_template = response.json()['body']
    except KeyError:
        raise click.ClickException("Invalid license name, use `license show` "
                                   "to get the all available licenses.")

    if license not in LICENSE_WITH_CONTEXT:
        echo(license_template, nl=False)
    else:
        context = re.findall(r'\[(\w+)\]', response.json()['body'])
        raw_context = {'fullname': fullname, 'year': year, 'email': email,
                       'project': project}
        user_context = {key: value for (key, value) in raw_context.items()
                        if value is not None}
        default_context = get_default_context()
        for (key, value) in user_context.items():
            default_context[key] = value
        for item in context:
            license_template = license_template.replace('[{0}]'.format(item),
                                                        default_context[item])
        echo(license_template, nl=False)


@cli.command()
@click.argument('license')
def context(license):
    """Show the default context for the license."""
    if license not in LICENSE_WITH_CONTEXT:
        echo("Just use it, there's no context for the license.")
    else:
        response = requests.get(
            'https://api.github.com/licenses/{0}'.format(license),
            headers={'accept': 'application/vnd.github.drax-preview+json'})
        context = re.findall(r'\[(\w+)\]', response.json()['body'])
        default_context = get_default_context()
        echo('The template has following defaults:')
        for item in context:
            echo('\t{0}: {1}'.format(item, default_context[item]))
        echo('You can overwrite them at your ease.')
